README.md
# simpleRgenius
A Simple(R) Fork of {rgenius}

Forked from: https://github.com/AlbertoAlmuinha/rgenius

Simpler functions and maintanence to perform similar functions as the original {rgenius}

Not intended for broad audience development


NEWS
Initial commit: version 0.0.1

R/get_lyrics.R
#' @title Get Song Lyrics
#' 
#' @description Get song lyrics and parts from a song or many songs by
#' an artist
#' 
#' @param artist_name Character (length = 1).
#' Uses the artist name to get correct lyrics.
#' The "primary artist" will be used to identify the
#' appropriate artist
#' 
#' @param song_names Character (vector).
#' Name or names of songs to get lyrics for (narrows down by \code{artist_name})
#' 
#' @param urls Character (vector).
#' URLs for the songs (rather than using artist and song names).
#' Skips search for proper information and provides exactly the URL
#' to the lyrics desired
#' 
#' @param verbose Boolean (length = 1).
#' Whether progress on which song lyrics are being obtained
#' should be printed.
#' Defaults to \code{TRUE}
#' 
#' @return Returns a data frame with artist, song name,
#' part of the song (e.g., verse, chorus), and lyric for each part
#' 
#' @examples
#' \dontrun{
#' get_lyrics(
#'   artist_name = "The Beatles",
#'   song_names = c(
#'     "Lucy in the Sky with Diamonds",
#'     "Magical Mystery Tour"
#'   )
#' )}
#' 
#' @author Alberto Almuina and Alexander P. Christensen <alexpaulchristensen@gmail.com>
#' 
#' @importFrom magrittr %>%
#' 
#' @export
#' 
# Gets all lyrics ----
# Updated 14.08.2023
get_lyrics <- function(artist_name, song_names, urls = NULL, verbose = TRUE)
{
  
  # Check if URLs are already provided
  if(is.null(urls)){
    
    # Check for verbose
    if(verbose){message("Fetching URLs...", appendLF = FALSE)}
    
    # Loop over song names
    SONG_URLS <- sapply(
      song_names, function(song_name){
        
        # Get top hits
        top_hits <- search_information(song_name, type = "song")
        
        # Set names to lower
        song_name <- song_name
        top_hits$song <- top_hits$song
        
        # Update song names
        top_hits$song <- convert_HTML_characters(top_hits$song)
        
        # Get top hits with artist name
        top_hits <- top_hits[top_hits$primary_artist == artist_name,]
        
        # Get top hits with song name
        top_hits <- top_hits[
          grepl(tolower(grepl_escapes(song_name)), tolower(top_hits$song)),
        ]
        
        # Check for exact match
        exact_match <- grepl(song_name, top_hits$song)
        
        # Use exact match (if possible)
        if(sum(exact_match) == 1){
          return(top_hits[exact_match, "url"])
        }else{ # Return top hit with artist information
          return(top_hits[which.max(top_hits$page_views), "url"])
        }
        
      }
    )
    
    # Return done
    message("done")
    
  }else{SONG_URLS <- urls}
  
  # Check for verbose
  if(verbose){message("Getting lyrics for...")}
  
  # Set en-dash
  en_dash <- rawToChar(as.raw(c(0xE2, 0x80, 0x93)))
  
  # Get all lyrics
  all_lyrics <- lapply(SONG_URLS, function(URL){
    
    # Get HTML output
    HTML_output <- rvest::read_html(URL)
    
    # Get artist and song information
    ARTIST_SONG <- trimws(
      gsub("\\|.*", "", HTML_output %>% rvest::html_nodes('title') %>% rvest::html_text())
    )
    ARTIST <- trimws(gsub(paste0(en_dash, ".*"), "", ARTIST_SONG))
    SONG <- trimws(gsub("Lyrics", "", gsub(paste0(".*", en_dash), "", ARTIST_SONG)))
    
    # Print message about getting lyrics
    if(verbose){message(SONG)}
    
    # Extract lyrics
    LYRICS <- HTML_output %>% rvest::html_nodes(xpath = '//*[@data-lyrics-container="true"]') %>% rvest::html_text()
    
    # Get parts from inside brackets
    parts <- stringr::str_extract_all(LYRICS, "\\[.*?\\]") %>% unlist()
    
    # Remove instrumental parts
    parts <- parts[!grepl("Instrumental", parts)]
    parts <- parts[!grepl("Non-Lyrical Break", parts)]
    
    # Extract the text outside brackets
    outside_brackets <- stringr::str_split(LYRICS, "\\[.*?\\]") %>% unlist()
    outside_brackets <- outside_brackets[outside_brackets != ""]
    
    # Get lyrics that are split
    lyrics <- split_words(outside_brackets)
    
    # Remove "( " space
    lyrics <- gsub("\\( ", "\\(", lyrics)
    
    # Return data frame
    return(
      data.frame(
        Artist = ARTIST,
        Song = SONG,
        Part = parts,
        Lyric = lyrics
      )
    )
    
    
  })
  
  # Get all lyrics as a data frame
  all_lyrics <- do.call(rbind.data.frame, all_lyrics)
  row.names(all_lyrics) <- NULL
  
  # Trim all lyrics one last time
  all_lyrics$Lyric <- trimws(all_lyrics$Lyric)
  all_lyrics$Lyric <- gsub("  ", " ", all_lyrics$Lyric)
  
  # Substitute HTML characters
  all_lyrics$Song <- convert_HTML_characters(all_lyrics$Song)
  
  # Return results as data frame
  return(all_lyrics)
  
}




R/get_discography.R
#' @title Get Discography
#' 
#' @description Gets full information for an artist
#' 
#' @param name Character (length = 1).
#' An artist's name
#' 
#' @return
#' Returns a data frame with the artists, song names,
#' Genius ID, Genius URL, primary artist, primary artist Genius ID,
#' and primary artist Genius URL
#' 
#' @author Alberto Almuina and Alexander P. Christensen <alexpaulchristensen@gmail.com>
#' 
#' @examples
#' \dontrun{
#' # Search for "The Beatles"
#' get_discography("The Beatles")}
#' 
#' @export
# Search information based on name
# Updated 08.08.2023
get_discography <- function(name)
{
  
  # Fetch content
  available_content <- httr::content(
    httr::RETRY(
      'GET', url = "https://api.genius.com/search",
      httr::user_agent("Mozilla/5.0"),
      query = list(
        q = name, page = 1, per_page = 50,
        access_token = Sys.getenv('GENIUS_API_TOKEN')
      ), quiet = TRUE
    )
  )
  
  # Get initial information
  initial_information <- get_information(
    available_content$response$hits, name, "artist"
  )
  
  # Get artist id
  ARTIST_ID <- initial_information$primary_artist_id[
    which.max(initial_information$primary_artist == name)
  ] 
  
  # Initialize search page
  page <- 1
  
  # Initialize result list
  result_list <- list()
  
  # Loop over search pages
  while(TRUE){
    
    # Fetch content
    available_content <- httr::content(
      httr::RETRY(
        'GET', url = paste0(
          "https://api.genius.com/artists/", ARTIST_ID, "/songs"
        ),
        httr::user_agent("Mozilla/5.0"),
        query = list(
          q = name, page = page, per_page = 50,
          access_token = Sys.getenv('GENIUS_API_TOKEN')
        ), quiet = TRUE
      )
    )
    
    # Check for failed search
    failed_search(available_content)
    
    # Get information
    result_list[[page]] <- get_information(
      available_content$response$songs, "artist"
    )
    
    # Check for next page
    if(is.null(available_content$response$next_page)){
      break
    }else{ 
      
      # Increase page
      page <- page + 1
      
      # Update page
      cat("\r Searching on page: ", page, rep(" ", 10))
      
    }
    
  }
  
  # Return unique results
  return( 
    unique(
      do.call(rbind.data.frame, result_list), MARGIN = 1
    )
  )
  
}







R/search_information.R
#' @title Search By Artist or Song for Information
#' 
#' @description Get basic Genius information for an artist or song title
#' 
#' @param name Character (length = 1).
#' Either an artist or a song
#' 
#' @param type Character (length = 1).
#' Specify whether \code{name} is an artist or a song
#' 
#' @return
#' Returns a data frame with the artists, song names,
#' Genius ID, Genius URL, primary artist, primary artist Genius ID,
#' and primary artist Genius URL
#' 
#' @author Alberto Almuina and Alexander P. Christensen <alexpaulchristensen@gmail.com>
#' 
#' @examples
#' \dontrun{
#' # Search for "The Beatles"
#' search_information("The Beatles", type = "artist")}
#' 
#' @export
# Search information based on name
# Updated 08.08.2023
search_information <- function(name, type = c("artist", "song"))
{
  
  # Fetch content
  available_content <- httr::content(
    httr::RETRY(
      'GET', url = "https://api.genius.com/search",
      httr::user_agent("Mozilla/5.0"),
      query = list(
        q = name, page = 1, per_page = 50,
        access_token = Sys.getenv('GENIUS_API_TOKEN')
      ), quiet = TRUE
    )
  )
  
  # Check for failed search
  failed_search(available_content)
  
  # Return results
  return(get_information(available_content$response$hits, name, type))
  
}


R/helpers.R
#%%%%%%%%%%%%%%%%%%%%%%
# *APPLY FUNCTIONS ----
#%%%%%%%%%%%%%%%%%%%%%%

# These functions are merely wrappers over the `*apply`
# family to execute specific tasks that are often repeatedly
# done. Their purpose are to provide slightly cleaner code.

# These `*vapply` are specific versions of `vapply` that
# pre-specify the output. These versions are generally
# simpler (in function) and marginally faster than 
# the more commonly used `sapply`. They are more strict
# in that they require knowing the output format (e.g., numeric)

#' @noRd
# Character vector/matrix output apply ----
# Updated 06.07.2023
cvapply <- function(X, FUN, ..., LENGTH = 1, USE.NAMES = TRUE)
{
  return(vapply(X = X, FUN = FUN, FUN.VALUE = character(LENGTH), ..., USE.NAMES = USE.NAMES))
}

#' @noRd
# Logical vector/matrix output apply ----
# Updated 06.07.2023
lvapply <- function(X, FUN, ..., LENGTH = 1, USE.NAMES = TRUE)
{
  return(vapply(X = X, FUN = FUN, FUN.VALUE = logical(LENGTH), ..., USE.NAMES = USE.NAMES))
}

#' @noRd
# Numeric vector/matrix output apply ----
# Updated 06.07.2023
nvapply <- function(X, FUN, ..., LENGTH = 1, USE.NAMES = TRUE)
{
  return(vapply(X = X, FUN = FUN, FUN.VALUE = numeric(LENGTH), ..., USE.NAMES = USE.NAMES))
}

#' @noRd
# Unlist `lapply` ----
# Updated 14.07.2023
ulapply <- function(X, FUN, ..., recursive = TRUE)
{
  return(unlist(lapply(X, FUN, ...), recursive = recursive))
}


#%%%%%%%%%%%%%%%%%%
# WEB SCRAPING ----
#%%%%%%%%%%%%%%%%%%

#' @noRd
# General error in failed search ----
# Updated 08.08.2023
# Check for meta status
failed_search <- function(available_content){
  
  if(available_content$meta$status != 200){
    stop(
      paste0(
        "Information on could not be found.",
        "\nIs spelling, punctuation, and spacing correct?"
      ), call. = FALSE
    )
  }
  
}

#' @noRd
# Get information function
# Updated 08.08.2023
get_information <- function(information, name, type)
{
  
  # Check for "result" in list name
  if("result" %in% names(information[[1]])){
    information <- lapply(information, function(x){x$result})
  }else if("response" %in% names(information[[1]])){
    information <- lapply(information, function(x){x$response})
  }
  
  # Remove incomplete lyrics
  keep_information <- lvapply(information, function(x){
    x$lyrics_state == "complete"
  })
  
  # Keep only complete lyrics
  information <- information[keep_information]
  
  # Get information
  if(type == "artist"){
    
    # Get only primary artist
    index <- which(
      lvapply(
        information, function(x){
          x$primary_artist$name == name
        }
      )
    )
    
  }else if(type == "song"){
    
    # Otherwise, use all indices
    index <- seq_along(information)
    
  }
  
  # Loop over information to get values
  return( 
    unique( # Get only unique
      do.call(
        rbind.data.frame, # Make into a data frame
        lapply(information[index], function(x){
          
          return(
            data.frame(
              artists = x$artist_names,
              song = x$title,
              id = x$id, url = x$url,
              primary_artist = x$primary_artist$name,
              primary_artist_id = x$primary_artist$id,
              primary_artist_url = x$primary_artist$url,
              page_views = ifelse(
                "pageviews" %in% names(x$stats),
                x$stats$pageviews, 0
              )
            )
          )
          
        })
      ), MARGIN = 1
    )
  )
  
}

#' @noRd
# Function to split words from GPT-4
# Updated 08.08.2023
split_words <- function(text) {
  
  # Replace capital letters with a space and the capital letter
  text <- stringr::str_replace_all(text, "([A-Z])", " \\1")
  
  # Replace punctuation followed by a non-space character with the punctuation and a space,
  # but exclude contractions with 's or 't, and exclude dashes
  text <- stringr::str_replace_all(text, "(\\p{P}(?<![''-])(?<![''](s|t)))([^\\s])", "\\1 \\2")
  
  # Remove extra spaces that may have been added at the beginning
  text <- stringr::str_replace(text, "^\\s+", "")
  
  # Return text
  return(text)
  
}

#%%%%%%%%%%%%%%%%%%%%%
# TEXT PROCESSING ----
#%%%%%%%%%%%%%%%%%%%%%

#' @noRd
# Character conversion
# Updated 13.08.2023
convert_HTML_characters <- function(string)
{
  
  # List of characters
  en_dash <- rawToChar(as.raw(c(0xE2, 0x80, 0x93)))
  right_apostrophe <- rawToChar(as.raw(c(0xE2, 0x80, 0x99)))
  
  # Perform conversion
  return(
    gsub(
      en_dash, "-", gsub(
        right_apostrophe, "'", string
      )
    )
  )
  
  
}

#' @noRd
# `grepl` escapes
# Updated 13.08.2023
grepl_escapes <- function(string)
{
  
  # List of characters
  grepl_characters <- c(
    "\\", "^", "$", ".", "|",
    "?", "*", "+", "(", ")",
    "[", "]", "{", "}"
  )
  
  # Loop over characters
  for(char in grepl_characters){
    string <- gsub(
      paste0("\\", char),
      paste0("\\\\", char),
      string
    )
  }
  
  # Return string
  return(string)
  
}










